### PR TITLE
FIX-1013: remove detail::value type

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
We would like `eve::value_type` to have ranges semantics and use `eve::element_type` for maybe going from wide to scalar.

This required:
a few concept changes
removing sign_of, since I couldn't make it work for wide in meta and it was hacked anyways. It's used only in one place, so I just implemented a version there.
A lot of small changes.

You can review by commits if you want to.

### CODE REVIEW

File: include/eve/module/proba/regular/impl/lognormal_distribution.hpp [hunk 1]
```diff
@@ -270,7 +270,7 @@ namespace eve
         return median(d)*exp(-sqr(d.s));
       else
       {
-        value_type inve = I(0.36787944117144232159552377016146086744581113103176);
+        auto inve = I(0.36787944117144232159552377016146086744581113103176);
```
Comment (jfalcou) [inline include/eve/module/proba/regular/impl/lognormal_distribution.hpp:273]: this value_type is a local typedef, I think it should stay the way it was.

File: include/eve/module/real/algorithm/function/regular/generic/reduce.hpp [hunk 2]
```diff
@@ -54,26 +54,26 @@ namespace eve::detail
     return op(s, v);
   }
 
-  template<real_scalar_value T, typename N, typename Callable>
+  template<scalar_value T, typename N, typename Callable>
```
Comment (jfalcou) [inline include/eve/module/real/algorithm/function/regular/generic/reduce.hpp:57]: Why this change ? I guess tests on tuple-like types ?

File: include/eve/traits/as_integer.hpp [hunk 3]
```diff
@@ -9,15 +9,26 @@
...
 #include <eve/forward.hpp>
 #include <eve/detail/meta.hpp>
+#include <eve/traits/element_type.hpp>
 
 namespace eve
 {
   namespace detail
   {
-    template<typename T> struct sign_of<logical<T>> : sign_of<T> {};
+    template<typename T>
```
Comment (jfalcou) [inline include/eve/traits/as_integer.hpp:18]: if it works here, why not put it back in meta.hpp as sign_of<T> ?

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/real/algorithm/function/regular/generic/reduce.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/algorithm/function/regular/generic/reduce.hpp:57]: Well reduce works on bundles. And somehow concepts restrictions here didn't work.

File: include/eve/traits/as_integer.hpp (hunk 3 above)
Comment (DenisYaroshevskiy) [reply include/eve/traits/as_integer.hpp:18]: I can't use `element_type_t` in `meta` - it depends on `forward` which depends on `meta`.
Also `meta` does not know about `eve` types - this is why you had the hack here for logicals.

File: include/eve/module/proba/regular/impl/lognormal_distribution.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/proba/regular/impl/lognormal_distribution.hpp:273]: It's not a local typedef, it's likely dead/untested code.

There is no `value_type` defined. It was just never compiled.
And now there's no `value_type` struct, so even in a not-instantiated form it's not working.

File: include/eve/module/real/algorithm/function/regular/generic/reduce.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/real/algorithm/function/regular/generic/reduce.hpp:57]: Good

File: include/eve/traits/as_integer.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/traits/as_integer.hpp:18]: OK good

**Review by jfalcou — APPROVED** [review]
